File: LeetCode/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int currMax = 0, finalMax = 0;
        int n = prices.length;
        for (int i = 1; i < n; i++) {
            currMax = Math.max(0, currMax += prices[i] - prices[i-1]);
            finalMax = Math.max(finalMax, currMax);
        }
        return finalMax;
    }
}

File: LeetCode/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxPrice, minPrice = 0, float('inf')
        for i in range(len(prices)):
            minPrice = min(minPrice, prices[i])
            maxPrice = max(maxPrice, prices[i] - minPrice)
        return maxPrice

# class Solution:
#     def maxProfit(self, prices: List[int]) -> int:
#         currMax = 0
#         finalMax = 0
#         n = len(prices)
#         for i in range(1, n):
#             currMax = max(0, currMax + prices[i] - prices[i-1])
#             finalMax = max(finalMax, currMax)
#         return finalMax